### PR TITLE
Classes to be compiled should only include classes with sources

### LINKED ISSUE #13767: Gradle incremental annotation processing fails with "Attempt to recreate a file" when a new file is included
Currently when a new file is added to a project the `Filer` object used by Gradle is somehow in an illegal state which results in compilation errors when an annotation processor tries to create a file. This is manifesting in Micronaut's support for incremental compilation in Gradle. For example:

```
> Task :compileTestJava FAILED
Note: Creating bean classes for 2 type elements
error: Unexpected error: Attempt to recreate a file for type incremental.$FooControllerTestDefinition
error: Unexpected error: Attempt to recreate a file for type incremental.$IncrementalTestDefinition
2 error
```

The compilation error can be resolved by simply running `./gradlew cTJ` again so it seems that Gradle's `Filer` object is in an illegal state at the daemon level. I have reviewed what Micronaut is doing and we are definitely not calling `file.createClassFile(..)` multiple times which is what would trigger this error under normal circumstances and since the problem only manifests when incremental compilation is enabled this seems to be a Gradle bug.

### Expected Behavior

Incremental compilation should consistently work regardless if new files are added.

### Current Behavior

Incremental compilation fails when a new file is added.

### Context

Micronaut users using Gradle incremental compilation receive a hard to understand error message when adding a new source file.

### Steps to Reproduce 

```
$ sdk install micronaut 2.0.0
$ sdk use micronaut 2.0.0
$ mn create-app incremental
$ cd incremental
$ mn create-controller foo
$ mn create-bean bar
$ mn create-bean baz
$ ./gradlew cTJ
$ mn create-bean other
$ ./gradlew cTJ

> Task :compileJava
Note: Creating bean classes for 1 type elements

> Task :compileTestJava FAILED
Note: Creating bean classes for 2 type elements
error: Unexpected error: Attempt to recreate a file for type incremental.$FooControllerTestDefinition
error: Unexpected error: Attempt to recreate a file for type incremental.$IncrementalTestDefinition
2 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

### Your Environment

```
------------------------------------------------------------
Gradle 6.5
------------------------------------------------------------

Build time:   2020-06-02 20:46:21 UTC
Revision:     a27f41e4ae5e8a41ab9b19f8dd6d86d7b384dad4

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          14 (Oracle Corporation 14+36-1461)
OS:           Mac OS X 10.15.5 x86_64
```

### PR DESCRIPTION
**Pull request by melix**
It's possible that an annotation processor directly generates bytecode.
In this case, the generated classes shouldn't be part of the classes
to be recompiled even if they are annotated by an annotation recognized
by an annotation processor.

Fixes #13767

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
The only thing I'm a bit worried about is that I had to switch to a concrete type (`ClassSymbol`) to figure out if there's an associated source file. This doesn't seem to be a problem but it's kind of weak...

**Comment by graemerocher** [comment]
@melix I guess you could add a soft dependency on the JDK internal classes via reflection, although I'm not sure if that is better

**Comment by graemerocher** [comment]
Any other steps need to get this PR into ready to merge state?

**Comment by ljacomet** [comment]
I just took a look as there are tests failures related to the use of `com.sun.tools.javac.util.Name` in tests which does not seem available in Java 8.
I believe this will have to wait until we can figure out if the changes and the use of non public API is OK or not.

Note that this will be done before Gradle 6.7 goes out but not right now, summer time impacting availability.

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@ljacomet PTAL

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java [hunk 1]
```diff
@@ -162,7 +161,8 @@ private Processor decorateForIncrementalProcessing(Processor processor, Incremen
     }
 
     private Processor decorateForTimeTracking(Processor processor, AnnotationProcessorResult processorResult) {
-        return new TimeTrackingProcessor(processor, processorResult);
+        return processor;
+        //return new TimeTrackingProcessor(processor, processorResult);
```
Comment (ljacomet) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java:165]: Is that a real change or a forgotten local modification?
Either way, needs to be cleaned up ;-)

File: subprojects/language-java/language-java.gradle.kts [hunk 2]
```diff
@@ -60,6 +60,11 @@ dependencies {
     }
     integTestDistributionRuntimeOnly(project(":distributionsCore"))
     crossVersionTestDistributionRuntimeOnly(project(":distributionsBasics"))
+
+    buildJvms.whenTestingWithEarlierThan(JavaVersion.VERSION_1_9) {
```
Comment (ljacomet) [inline subprojects/language-java/language-java.gradle.kts:64]: 👍

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java (hunk 1 above)
Comment (melix) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java:165]: This is indeed not intended, good catch!

**Comment by melix** [comment]
> Btw how will this behave on JDKs that are not OpenJDK based?

I tested with IBM JDK 8 and it works.

**Pull request merged**